### DISCUSSION TITLE
Directus Cloud & Cloudinary

### DISCUSSION THREAD

**Discussion by sdevogel** [directus/directus#14649] [Archive]
Hi, I was wondering if it is possible with directus cloud to upload all media (images & videos) to cloudinary and be served from there? 

Is it possible at all to have storage adapters when hosted on directus cloud or is this only possible with self hosting?

Thanks in advance!

**Comment by rijkvanzanten** [comment]
Heya! Directus doesn't currently support Cloudinary in general. Cloud relies on it's own file storage and CDN for assets, so there's no third party storage adapters available on cloud yet 👍🏻

**Reply by SalahAdDin** [reply]
> Heya! Directus doesn't currently support Cloudinary in general. Cloud relies on its own file storage and CDN for assets, so there are no third-party storage adapters available on the cloud yet 👍🏻

Does it mean they could be supported in the future?

**Reply by rijkvanzanten** [reply]
Potentially. Cloudinary is it's own DAM, so we'd have to figure out the two-way binding logic first, as any changes made in Cloudinary directly have to be synced back to Directus

**Reply by SalahAdDin** [reply]
> Potentially. Cloudinary is its own DAM, so we'd have to figure out the two-way binding logic first, as any changes made in Cloudinary directly have to be synced back to Directus

Is it that so complex? the idea is to control the images from **Directus** but store them in **Cloundinary**.

**Reply by rijkvanzanten** [reply]
Yes it's very complex, as both **Directus** and **Cloudinary** will change the files at the same time. **Directus** will not be aware of the changes that **Cloudinary** made, so you end up with a **corrupt** file library in **Directus**.

**Reply by SalahAdDin** [reply]
> Yes it's very complex, as both **Directus** and **Cloudinary** will change the files at the same time. **Directus** will not be aware of the changes that **Cloudinary** made, so you end up with a **corrupt** file library in **Directus**.

Can **Directus** receive any response or notification from **Cloudinary**?

**Reply by rijkvanzanten** [reply]
> Can Directus receive any response or notification from Cloudinary?

That's where it gets very complex 🙊

**Reply by SalahAdDin** [reply]
Sad to hear that.

**Reply by sharkfin009** [reply]
@rijkvanzanten In my case I have a self hosted Directus app and am trying to use the cloudinary driver mentioned in the [docs](https://docs.directus.io/self-hosted/config-options.html#cloudinary-cloudinary).

How is that supposed to work exactly? I am getting a "Driver cloudinary is not supported" error atm.

STORAGE_LOCATIONS=cloudinary
STORAGE_CLOUDINARY_DRIVER=cloudinary
STORAGE_CLOUDINARY_ROOT=
STORAGE_CLOUDINARY_HEALTHCHECK_THRESHOLD=750
STORAGE_CLOUDINARY_CLOUD_NAME=mateas-pics
STORAGE_CLOUDINARY_API_KEY=<key>
STORAGE_CLOUDINARY_API_SECRET=<secret>
STORAGE_CLOUDINARY_ACCESS_MODE=authenticated

Here is my .env
Thanks for any help!

**Reply by paescuj** [reply]
@sharkfin009 Please make sure you're running on the latest version of Directus (currently v9.22.4).

**Reply by sharkfin009** [reply]
Hi @paescuj ! thanks, just did that and that fixed it ! but now its popping this:
![Screen Shot 2023-02-07 at 16 09 38](https://user-images.githubusercontent.com/30635121/217268087-3cb027b6-8e70-4fbd-bc30-a4d535cefe74.png)

**Reply by sharkfin009** [reply]
[14:09:30.231] ERROR: Readable.fromWeb is not a function
err: {
"type": "TypeError",
"message": "Readable.fromWeb is not a function",
"stack":
TypeError: Readable.fromWeb is not a function
at DriverCloudinary.read (file:///app/node_modules/@directus/storage-driver-cloudinary/dist/index.js:102:25)
at runMicrotasks (<anonymous>)
at processTicksAndRejections (node:internal/process/task_queues:96:5)
at async FilesService.uploadOne (/app/node_modules/directus/dist/services/files.js:99:28)
at async Multipart.<anonymous> (/app/node_modules/directus/dist/controllers/files.js:78:32)
}

**Reply by paescuj** [reply]
I assume you're not self-hosting using the Directus Docker image (https://docs.directus.io/self-hosted/docker-guide.html), right? In that case also make sure to use the active LTS version of Node.js (currently v18).

**Reply by sharkfin009** [reply]
ok will check the node version. I'm using railway's starter, which probably uses that image?

**Reply by sharkfin009** [reply]
ah, yes 16.12.0

**Reply by sharkfin009** [reply]
fantastic , files are being uploaded to cloudinary. Not showing in the directus dash though

**Reply by paescuj** [reply]
> fantastic , files are being uploaded to cloudinary. Not showing in the directus dash though

Okay, great! Please double check your configuration. If the problem still persists you might want to open a corresponding bug report: https://github.com/directus/directus/issues

**Reply by sharkfin009** [reply]
Sorted now ...ENV  access mode value must be public

**Reply by SalahAdDin** [reply]
How did you setup it?

**Reply by sharkfin009** [reply]
STORAGE_LOCATIONS=cloudinary
STORAGE_CLOUDINARY_DRIVER=cloudinary
STORAGE_CLOUDINARY_ROOT=
STORAGE_CLOUDINARY_HEALTHCHECK_THRESHOLD=750
STORAGE_CLOUDINARY_CLOUD_NAME=mateas-pics
STORAGE_CLOUDINARY_API_KEY=(key)
STORAGE_CLOUDINARY_API_SECRET=(secret)
STORAGE_CLOUDINARY_ACCESS_MODE=public
 
Basically like this, with the latest directus and node >=18.0.0 in the package.json

**Reply by sharkfin009** [reply]
you can create new versions of your pics (on cloudinary) for each layout, using the directus transformation presets. I can help if you want... so then cloudinary just gives us a bucket of files served via CDN, with transformations managed by directus. Pretty happy with it

**Reply by sharkfin009** [reply]
now suddenly the img requests are getting 403's . Happened earlier and went away, now its back ;(

**Reply by sharkfin009** [reply]
hang on these are 400s, must be me somehow lol

**Reply by sharkfin009** [reply]
@paescuj if you wouldnt mind, do you have any thoughts on why 403's would be coming back when I apply any transformation parameters or a preset?

**Reply by paescuj** [reply]
> @paescuj if you wouldnt mind, do you have any thoughts on why 403's would be coming back when I apply any transformation parameters or a preset?

Hard to tell. To see what the problem might be, more information is needed. Probably best to open a separate [discussion](https://github.com/directus/directus/discussions) for this or if you are sure that it's a bug, a [bug report](https://github.com/directus/directus/issues/new/choose).

**Reply by jbmolle** [reply]
We also have the 403 problem. It's coming from Cloudinary Admin API when we reach the 500 requests / hour limit.
When we have too many images, Directus is probably calling the API too much. What would be a way to solve this?
